### ISSUE TITLE
[RFC]: Expert-granular MoE residency — drive UVA offload from EPLB's per-expert load statistics

### ISSUE THREAD

**Issue by Mohammad-Nabulsi** [vllm-project/vllm#57794]
### Motivation.

vLLM can already offload MoE expert weights to host memory, and it already knows which experts are hot. It cannot yet use the second fact to drive the first.

What exists on `main` today:

* `PrefetchOffloadConfig.offload_group_size` / `offload_num_in_group` — choose **which layers** to offload.
* `UVAOffloadConfig.cpu_offload_params` — choose **which parameters** to offload, by name segment. Its own docstring uses `"mlp.experts.w2_weight"` as the example.
* `offload_backend="uva"`, with `is_uva_available()` and `get_accelerator_view_from_cpu_tensor()` — zero-copy access to pinned host memory.
* `EPLBConfig.window_size` / `step_interval` and EPLB's `expert_load_window` — a rolling record of **observed per-expert load**, already used to periodically rearrange expert *placement* across ranks.

The gap is that `cpu_offload_params` matches whole tensors. It moves `w2_weight` for *all* experts in a layer, or none of them. There is no way to express "this fraction of experts is cold under current traffic; demote them and keep the rest resident", even though the statistics that identify them are already being collected, and the cadence for acting on them already exists.

For a wide-MoE model this is the difference between offload being unusable and being the thing that makes the model fit. Expert weights dominate, routing is heavily skewed in practice, and the cold tail is large and fairly stable over a window. Offloading whole expert tensors costs far too much; offloading the cold tail costs comparatively little and frees VRAM that is worth more as KV cache.

### Proposed Change.

Reuse EPLB's existing loop. It already answers "which experts are cold" and "when do we act on that". This proposal adds only "and therefore where does each expert live".

**1. A fraction, not a count.** One knob:

```
cold_expert_fraction: float = 0.0   # 0.0 disables; e.g. 0.15
```

Per layer, `n_cold = floor(num_local_experts * cold_expert_fraction)`, so it is model- and TP-independent. A raw count is deliberately not offered — it does not transfer across models, and it is the kind of number that ends up hardcoded in a deployment and silently wrong after a reshard.

**2. Residency derived from the load window, on the existing cadence.** At each EPLB rebalance step, rank experts by `expert_load_window` and mark the coldest `n_cold` per layer as host-resident; the rest stay device-resident. This is deliberately the *same* trigger and the *same* data EPLB already uses to rearrange placement — residency rebalance alongside placement rebalance, not a second scheduler.

Between rebalance steps the split is **static**. That is the point; see rejected alternatives.

**3. A two-bank forward.** Run the stock expert forward once over the device bank and once over the host bank, and sum. Two `expert_map`s. Router top-k computed once and reused. Shared experts added on the device call only, so they are not double-counted.

**4. Correct split ordering.** The split must happen *after* `process_weights_after_loading`, because the weights must already carry the quant method's real repacked layout. The kernel/config must be rebuilt for **both** banks — one config shared across two banks silently runs cold weights against hot scales, which does not crash and produces plausible wrong output. This was a real bug we hit.

**5. The ranking policy is a hook.** `expert_load_window` is the default input, but the selection itself should be pluggable so deployments can substitute their own signal. I am not proposing to upstream a particular ranker.

### Rejected alternatives.

**Per-token swap-on-demand with prefetch.** The obvious design — keep a small resident cache and fetch an expert from host when a token routes to it — does not work, and it is worth writing down why so nobody re-derives it.

MoE routing is data-dependent. The expert set for layer *L* is known only after computing layer *L*'s router logits, which requires layer *L−1*'s output. There is no lookahead window to prefetch into: by the time the requirement is known, the weights are needed immediately, on the critical path.

The cost, from our measurements (1 × H100 80 GB, 2.53 MB per expert per layer, 48 layers, 10 active experts/token):

| | |
|---|---|
| all-resident decode | 45.7 tok/s → 21.9 ms/token → ~456 µs/layer |
| measured host gather, one cold expert | 63.2 µs = **14% of one layer's budget** |
| if one of ten active experts per layer is cold | +3.0 ms/token → decode falls to ~40 tok/s |

And that is for a *single* cold expert per layer. Swapping also means mutating the device bank during inference, which invalidates CUDA-graph captures and changes `expert_map` in flight.

Periodic re-derivation gets the adaptivity without any of this: the transfer cost is paid once per rebalance interval rather than once per token, which amortizes it to nothing.

**Tiered hot/warm/cold with threshold rules.** Considered and dropped. We cannot justify specific tier thresholds with the data we have, and a fraction plus a pluggable policy spans the same space with one parameter instead of several.

### Feedback Period.

Two weeks.

### Any Other Things.

**The honest version of our measurements.**

Setup: 1 × H100 80 GB, a 48-layer MoE with 512 experts/layer and 10 active per token, 88 experts offloaded per layer (17.2%, the static equivalent of `cold_expert_fraction=0.17`). The checkpoint is assembled (compressed-tensors INT4 body with an FP8 embedding), not a published one — so these numbers are **not independently reproducible**, which is a real limitation of this evidence.

| | all resident | 88/512 offloaded |
|---|---|---|
| model memory at load | 68.18 GiB | 58.52 GiB (−9.66 GiB) |
| GPU KV cache | 144,725 tok | 522,878 tok |
| decode @ concurrency 1 | 45.7 tok/s | 37.1 tok/s (−19%) |
| latency @ concurrency 4 | 7.65 s | 3.67 s |

Four caveats a reviewer should have up front:

1. **The concurrency-4 improvement is not cheaper compute.** It is the larger KV pool avoiding preemption: `num_preemptions_total` went 0 → 3 in the all-resident run and stayed 0 throughout the offloaded one. Offload does not make decode faster; it stopped the scheduler thrashing.

2. **The cold-expert cost is 2–4× what the naive model predicts, and I cannot explain it.** Naive estimate 1.15–2.15 ms/token at the measured cold-hit fraction; observed ≈5.1 ms/token at concurrency 1. Unverified candidates: PCIe contention with other host-resident tensors streaming each step, ~48 extra kernel launches per decode step, hot-bank geometry change. **This is the single biggest open question** and the main thing I would want help with. EPLB's per-expert counters look like the right instrument for attacking it.

3. **Only 0 and 88 offloaded experts have ever been latency-measured.** Every other fraction is extrapolation sitting on top of the gap in (2).

4. **9.66 GiB is measured at load; the arithmetic prediction was 9.97 GiB.** The 3.1% gap is real and unexplained.

Net: the memory saving is solid and the KV headroom it buys is large. The latency story is workload-dependent — a loss at low concurrency, a win at higher concurrency for tail-heavy traffic — and it rests on a cost model that does not yet close. I would rather agree the shape here before re-implementing a ~600-line change to the MoE forward path against a tree that has since been restructured.

Two adjacent findings from the same work, filed separately as PRs: #57792 and #57793.

**Comment by Mohammad-Nabulsi** [comment]
Substantially revised the proposal. Three changes, all narrowing it:

1. **A fraction replaces a fixed expert count.** `cold_expert_fraction`, floored per layer, so the knob is model- and TP-independent. The original framing carried a hardcoded count from one deployment, which does not transfer.

2. **Residency is now re-derived on EPLB's existing cadence** rather than being fixed at load. `EPLBConfig.window_size` / `step_interval` and `expert_load_window` already answer "which experts are cold" and "when do we act on it" — this reuses both rather than adding a second scheduler. It also makes the feature adaptive to traffic, which the original was not.

3. **Added a rejected-alternatives section.** Per-token swap-on-demand with prefetch is the design everyone proposes first, including me. It cannot work: routing is data-dependent, so layer L's expert set is unknown until layer L-1 has produced its output, leaving no lookahead to prefetch into. Our measured host-gather cost is 63.2 µs against a ~456 µs per-layer decode budget, so a single cold expert per layer costs ~14% of that budget and ~3.0 ms/token end to end. It would also mutate the device bank mid-inference, invalidating CUDA-graph captures. Writing this down so nobody re-derives it.

The measurements and their caveats are unchanged.

**Comment by jmeadlock** [comment]
Data for this RFC from a second model and a coherent-memory box, in case it helps settle the shape. Same mechanism as §3–4 here: split each routed-expert layer after `process_weights_after_loading` into a device bank and a host bank, rebuild the quant config/kernel for **each** bank (+1 to your note that sharing one config silently runs cold weights against hot scales), two `expert_map`s, one router top-k, sum. Static placement between re-derivations.

**Setup.** One DGX Station GB300 (B300 + Grace, NVLink-C2C), MiMo-V2.6-Pro-RL (69 MoE layers × 384 routed experts, native MXFP4, 527 GiB on disk), vLLM nightly `d05da62e`, `--moe-backend marlin`, `--offload-backend uva --cpu-offload-gb 320 --cpu-offload-params routed_experts.w13_weight routed_experts.w2_weight`, `VLLM_WEIGHT_OFFLOADING_DISABLE_PIN_MEMORY=1`, FULL_DECODE_ONLY graphs, TP1. Host bank is exact-size pinned memory read in place over C2C (no H2D copy). Stock `--cpu-offload-gb` put whole layers 1–48 in Grace and 49–69 in HBM. Out-of-tree experiment (post-load hook), not a PR.

**1. Ranking signal generalises if it is decode counts from real traffic.** Router selections from 900 real agent turns (3.58M prompt / 89K generated tokens), 630 used to rank, 270 held out. Held-out *decode* traffic served from HBM at the stock byte budget:

| placement | held-out decode traffic from HBM |
|---|---|
| stock `--cpu-offload-gb` (layer order) | 30.4% |
| ranked on train decode counts | 62.2% |
| oracle (held-out ranks itself) | 62.5% |
| ranked on prefill counts only | 54.9% |

So for your §5 hook: an EPLB-style load window works as the default if it can separate decode from prefill; prefill-heavy windows rank noticeably worse.

**2. Parity.** 16-layer truncation, stock vs split, greedy: 6/6 token-identical; per-layer residual relative drift grows to 5.0e-4 by layer 15; max |Δlogprob| on chosen tokens 0.067. Summation-order drift from the two partial sums, not bit-identity (cf. the fp32-finalize variant earlier in #38256, which was bit-identical on a different kernel).

**3. Throughput — the opposite sign of your H100 result at C1, because here the budget is held equal rather than freed.** Matched A/B, same recipe, only the split toggled, warm runs, tok/s:

| | stock (layer order) | split, same HBM bytes | split, +11 GiB hot, −11 GiB KV |
|---|---|---|---|
| decode after TTFT, 11.5k-token prompt | 31.2 | 36.3 | 37.4 |
| agent tool-call JSON | 29.5 | 35.7 | 36.6 |
| shell / structured | 28.6 / 30.0 | 35.4 / 34.8 | 36.4 / 35.9 |
| prose (fixture) | 30.3 | 31.6 | 32.7 |
| synthetic prose, C1 | 30.0 | 29.6–30.0 | 29.7–30.1 |
| C8 aggregate | — | 60.8 | 63.7 |
| GPU KV tokens | 490K | 514K | 302K |

(The same-budget split came out with 24K *more* KV tokens than stock; I have not isolated why — probably the freed full-size scale tensors — so treat it as a side observation, not a claim.) The first ranking used agent traffic only and cost synthetic prose ~5% at C1; mixing a prose-leaning histogram in at 0.25 weight removed that at a 1.2-point coverage cost. The ranking corpus matters more than the split mechanics.

**4. On your open question (2), cold cost 2–4× the naive model.** A torch-profiler trace of 64 C1 decode tokens on the split lane: 28.8 ms/token, GPU busy 96% (not launch-bound). Marlin routed experts 17.3 ms/token, of which the hot bank is 3.4 ms and the cold bank 14.0 ms. The cold bank carries ~39% of routed traffic and ~80% of expert time: **~6× the cost per unit of traffic** of the device bank, for weights read in place at ~340 GB/s over C2C against HBM. That ratio is much closer to the bandwidth ratio than to a per-launch cost, so on this hardware the extra ~69 launches/token from the second bank were not the problem (GPU idle is 1.05 ms/token total). That is only evidence about launch count on this box; on a PCIe host bank the 2–4× gap you see may have a different cause, and I haven't measured one.

**What I'd suggest for the design:** keep the placement static and boot-time as a first PR (it is where all of the win here is), with the ranking input as a file or hook rather than tied to EPLB's window. EPLB re-derivation can come second. Happy to run a candidate branch on this box against the numbers above.
